packages/wp-now/src/pool.ts

```diff
@@ -0,0 +1,261 @@
+ */
+const spawn = (pool) => {
+	const newInstances = new Set();
+
+	while (pool.maxJobs > 0 && pool.instanceInfo.size < pool.maxJobs) {
+		const info = new PoolInfo();
+		const instance = pool.spawner();
+		pool.instanceInfo.set(instance, info);
+		info.active = true;
+		newInstances.add(instance);
+	}
+
+	return newInstances;
+};
+
+/**
+ * PRIVATE
+ * Reaps children if they've passed the maxRequest count.
+ */
+const reap = (pool) => {
+	for (const [instance, info] of pool.instanceInfo) {
+		if (pool.maxRequests > 0 && info.requests >= pool.maxRequests) {
+			info.active = false;
+			pool.instanceInfo.delete(instance);
+			instance.then((unwrapped) => unwrapped.exit()).catch((error) => {});
+			continue;
+		}
+	}
+};
+
+/**
+ * PRIVATE
+ * Handle fatal errors gracefully.
+ */
+const fatal = (pool, instance, error) => {
+	console.error(error);
+
+	if (pool.instanceInfo.has(instance)) {
+		const info = pool.instanceInfo.get(instance);
+		info.active = false;
+		pool.instanceInfo.delete(instance);
+	}
+};
+
+/**
+ * PRIVATE
+ * Find the next available idle instance.
+ */
+const getIdleInstance = (pool) => {
+	const sorted = [...pool.instanceInfo].sort(
+		(a, b) => a[1].requests - b[1].requests
+	);
+
+	for (const [instance, info] of sorted) {
+		if (pool.running.has(instance)) {
+			continue;
+		}
+
+		if (!info.active) {
+			continue;
+		}
+		return instance;
+	}
+
+	return false;
+};
+
+/**
+ * Maintains and refreshes a list of php instances
+ * such that each one will only be fed X number of requests
+ * before being discarded and replaced.
+ */
+export class Pool {
+	instanceInfo = new Map(); // php => PoolInfo
+
+	spawner: () => Promise<any>; // Callback to create new instances.
+	maxRequests: number; // Max requests to feed each instance
+	maxJobs: number; // Max number of instances to maintain at once.
+
+	notifiers = new Map(); // Inverted promises to notify async code of backlogged item processed.
+	running = new Set(); // Set of busy PHP instances.
+	backlog = []; // Set of request callbacks waiting to be run.
+
+	constructor({
+		spawner = async (): Promise<any> => {},
+		maxRequests = 2000,
+		maxJobs = 5,
+	} = {}) {
+		this.spawner = spawner;
+		this.maxRequests = maxRequests;
+		this.maxJobs = maxJobs;
+		reap(this);
+		spawn(this);
+	}
+
+	/**
+	 * Queue up a callback that will make a request when an
+	 * instance becomes idle.
+	 */
+	async enqueue(item: (php: NodePHP) => Promise<any>) {
+		const idleInstance = getIdleInstance(this);
+
+		if (!idleInstance) {
+			// Defer the callback if we don't have an idle instance available.
+			this.backlog.push(item);
+
+			// Split a promise open so it can be accepted or
+			// rejected later when the item is processed.
+			const notifier = new Promise((accept, reject) =>
+				this.notifiers.set(item, { accept, reject })
+			);
+
+			// Return the notifier so async calling code
+			// can still respond correctly when the item
+			// is finally processed.
+			return notifier;
+		} else {
```

Comment:
Nit: No need for the `else`

Reply:
These nitpicks tell me you're looking at my code closely.

I appreciate it, and I'll go ahead and make that change.